dali/pipeline/operators/crop/crop.cu

```diff
@@ -13,23 +13,201 @@
+  const auto output = ws->Output<GPUBackend>(idx);
+  DALI_CALL((BatchedCrop<Out>(
+      input_ptrs_gpu_.template data<const uint8*>(),
+      input_strides_gpu_.template data<int>(),
+      batch_size_, crop_[0], crop_[1], C_, output_layout_,
+      output->template mutable_data<Out>(),
+      ws->stream())));
+}
+
+template<>
+template <typename Out>
+void Crop<GPUBackend>::ValidateHelper(TensorList<GPUBackend> *output) {
+  // Validate parameters
+  DALI_CALL(ValidateBatchedCrop(
+    input_ptrs_.template mutable_data<const uint8*>(),
+    input_strides_.template data<int>(),
+    batch_size_, crop_[0], crop_[1], C_,
+    output->template mutable_data<Out>()));
+}
+
 template<>
 void Crop<GPUBackend>::SetupSharedSampleParams(DeviceWorkspace *ws) {
-  CropCastPermute::SetupSharedSampleParams(ws);
+  const auto & input = ws->Input<GPUBackend>(0);
+
+  if (output_type_ == DALI_NO_TYPE)
+    output_type_ = input.type().id();
+
+  for (int i = 0; i < batch_size_; ++i)
+    SetupSharedSampleParams(ws, input.tensor_shape(i), i, i);
+}
+
+template<>
+void Crop<GPUBackend>::DataDependentSetup(DeviceWorkspace *ws, const int idx) {
+  auto &input = ws->Input<GPUBackend>(idx);
+  auto output = ws->Output<GPUBackend>(idx);
+  DALI_ENFORCE(IsType<uint8>(input.type()),
+               "Expected input data as uint8.");
+
+  DALITensorLayout outLayout;
+  const Dims out_shape = GetOutShape(input.GetLayout(), &outLayout);
+
+  std::vector<Dims> output_shape(batch_size_);
+  for (int i = 0; i < batch_size_; ++i) {
+    const auto input_shape = input.tensor_shape(i);
+    DALI_ENFORCE(input_shape.size() == 3,
+                 "Expects 3-dimensional image input.");
+
+    const int H = input_shape[0];
+    const int W = input_shape[1];
+
+    DALI_ENFORCE(H == per_sample_dimensions_[i].first &&
+                 W == per_sample_dimensions_[i].second,
+         "Corresponding images in different input sets need to have the same height and width");
+    const int C = input_shape[2];
+
+    DALI_ENFORCE(C == C_,
+                 "Input channel dimension does not match "
+                 "the output image type. Expected input with "
+                 + to_string(C_) + " channels, got " + to_string(C) + ".");
+
+    const int crop_y = per_sample_crop_[i].first;
+    const int crop_x = per_sample_crop_[i].second;
+
+    input_strides_.template mutable_data<int>()[i] = W*C;
+    crop_offsets_[i] = (crop_y * W + crop_x) * C;
+    output_shape[i] = out_shape;
+  }
+
+  output->Resize(output_shape);
+  output->SetLayout(outLayout);
+
+  // Calculate input pointers and copy to gpu
+  for (int i = 0; i < batch_size_; ++i) {
+    input_ptrs_.template mutable_data<const uint8*>()[i] =
+      input.template tensor<uint8>(i) + crop_offsets_[i];
+  }
+  input_ptrs_gpu_.Copy(input_ptrs_, ws->stream());
+  input_strides_gpu_.Copy(input_strides_, ws->stream());
+
+  // Validate
+  if (output_type_ == DALI_FLOAT) {
```

Comment:
Maybe we could just call ValidateHelper from RunHelper to avoid this if/else ladder?